zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,373 @@
+
+        mostRecentLogNeededIndex = i;
+        break;
+      }
+    }
+
+    return validateLogRange();
+  }
+
+  private boolean validateLogRange() {
+    Preconditions.checkState(oldestLogNeededIndex >= 0);
+    Preconditions.checkState(oldestLogNeededIndex < logs.size());
+    Preconditions.checkState(mostRecentLogNeededIndex >= 0);
+    Preconditions.checkState(mostRecentLogNeededIndex < logs.size());
+    Preconditions.checkState(oldestLogNeededIndex >= mostRecentLogNeededIndex);
+    Preconditions.checkState(snapNeededIndex >= 0);
+    Preconditions.checkState(snapNeededIndex < snaps.size());
+
+    BackupFileInfo snap = snaps.get(snapNeededIndex);
+    BackupFileInfo oldestLog = logs.get(oldestLogNeededIndex);
+    BackupFileInfo newestLog = logs.get(mostRecentLogNeededIndex);
+
+    if (oldestLog.getFileType() == BackupFileType.LOSTLOG) {
+      LOG.error("Could not find logs to make the snapshot '" + snap.getBackedUpFile()
+          + "' valid. Lost logs at " + logs.get(oldestLogNeededIndex).getZxidRange());
+      return false;
+    }
+
+    if (newestLog.getFileType() == BackupFileType.LOSTLOG) {
+      if (zxidToRestore == Long.MAX_VALUE && oldestLogNeededIndex != mostRecentLogNeededIndex) {
+        // When restoring to the latest, we can use the last valid log prior to lost log
+        // range.
+        mostRecentLogNeededIndex++;
+      } else {
+        LOG.error("Could not find logs to make the snapshot '" + snap.getBackedUpFile()
+            + "' valid. Lost logs at " + logs.get(mostRecentLogNeededIndex).getZxidRange() + ".");
+        return false;
+      }
+    }
+
+    Range<Long> fullRange = oldestLog.getZxidRange().span(newestLog.getZxidRange());
+
+    if (fullRange.lowerEndpoint() > snap.getZxidRange().lowerEndpoint()) {
+      LOG.error("Could not find logs to make snap '" + snap.getBackedUpFile()
+          + "' valid. Logs start at zxid " + ZxidUtils.zxidToString(fullRange.lowerEndpoint())
+          + ".");
+      return false;
+    }
+
+    if (fullRange.upperEndpoint() < snap.getZxidRange().upperEndpoint()) {
+      LOG.error("Could not find logs to make snap '" + snap.getBackedUpFile()
+          + "' valid. Logs end at zxid " + ZxidUtils.zxidToString(fullRange.upperEndpoint()) + ".");
+      return false;
+    }
+
+    if (zxidToRestore != Long.MAX_VALUE && fullRange.upperEndpoint() < zxidToRestore) {
+      LOG.error("Could not find logs to restore to zxid " + zxidToRestore + ". Logs end at zxid "
+          + ZxidUtils.zxidToString(fullRange.upperEndpoint()) + ".");
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Copy selected backup files from backup storage to a local restore temporary directory for further processing later
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @throws IOException
+   */
+  private void copyRestoredFilesToLocalTempDir(FileTxnSnapLog restoreTempSnapLog)
```

Comment:
Nit: `copyBackupFilesToLocalTempDir` sounds better?

I'm not sure if `restored` is the right way to describe these files because to me, `restored` means all processing/moving are complete (aka restoration is complete). In this case, we are referring to the backup files prior to any processing.